deb/s3apt.py

```diff
@@ -0,0 +1,279 @@
+        return
+
+    pkginfos = []
+    for obj in deb_objs:
+        print(obj.key)
+
+        pkginfo = get_cached_control_data(obj)
+        if obj.key.startswith(prefix):
+            filename = obj.key[len(prefix):]
+            pkginfo = pkginfo + "\n%s\n" % ("Filename: %s" % filename)
+        else:
+            pkginfo = pkginfo + "\n%s\n" % ("Filename: %s" % obj.key)
+
+        pkginfos.append(pkginfo)
+
+    package_index_obj = s3.Object(bucket_name=os.environ['BUCKET_NAME'], key=prefix + "Packages")
+    print("Writing package index: %s" % (str(package_index_obj)))
+    if os.environ['PUBLIC'] == 'True' :
+        acl = 'public-read'
+    else:
+        acl = 'private'
+    package_index_obj.put(Body="\n".join(sorted(pkginfos)), Metadata={'packages-hash': calcd_pkghash}, ACL=acl)
+
+    print("DONE REBUILDING PACKAGE INDEX")
+
+def calc_package_index_hash(deb_names):
+    """
+    Calculates a hash of all the given deb file names. This is deterministic so
+    we can use it for short-circuiting.
+    """
+
+    md5 = hashlib.md5()
+    md5.update("\n".join(sorted(deb_names)))
+    return md5.hexdigest()
+
+def get_package_index_hash(prefix):
+    """
+    Returns the md5 hash of the names of all the packages in the index. This can be used
+    to detect if all the packages are represented without having to load a control data cache
+    file for each package.can be used
+    to detect if all the packages are represented without having to load a control data cache
+    file for each package.
+    """
+    s3 = boto3.resource('s3')
+    try:
+        print("looking for existing Packages file: %sPackages" % prefix)
+        package_index_obj = s3.Object(bucket_name=os.environ['BUCKET_NAME'], key=prefix + 'Packages')
+        return package_index_obj.metadata.get('packages-hash', None)
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == '404':
+            return None
+        else:
+            raise(e)
+
+def build_release_file(prefix):
+    """
+    gets info from Package, get the sums and puts them into file
+    """
+    s3 = boto3.client('s3')
+    release_file = ""
+    s3.download_file(os.environ['BUCKET_NAME'], prefix + "Packages", '/tmp/Packages')
+    md5, sha1, sha256 = checksums("/tmp/Packages")
+
+    time = 'Date: ' + strftime("%a, %d %b %Y %X UTC", gmtime())
+    stat = os.stat("/tmp/Packages")
+    release_file = release_file +(time + '\nMD5sum:\n ' + md5)
+    for i in range(0,17-len(str(stat.st_size))):
+        release_file = release_file +(' ')
+    release_file = release_file +(str(stat.st_size) + ' Packages\nSHA1:\n '+sha1 )
+    for i in range(0,17-len(str(stat.st_size))):
+        release_file = release_file +(' ')
+    release_file = release_file +(str(stat.st_size) + ' Packages\nSHA256:\n '+sha256 )
+    for i in range(0,17-len(str(stat.st_size))):
+        release_file = release_file +(' ')
+    release_file = release_file +(str(stat.st_size) + ' Packages')
```

Comment:
All the repetitions of `release_file` make this really hard to read.

* `release_file = release_file` could be replaced with `release_file +=`
* `str(stat.st_size) + ' Packages\nSHA1:\n '+sha1` could be replaced with `"%d Packages\nSHA1:\n %s" % stat.st_size, sha1`